### PR TITLE
Output html files for editor apps (widget wrapper + settings)

### PR DESCRIPTION
**Pull request by yavorsky**
Screen Shot 2020-02-25 at 12 48 28 PM

### 🔦 Summary
This is a follow-up for https://github.com/wix/yoshi/issues/2013

Currently, I don't have strong opinion about generating html files instead of serving own app on prod. But this is an optional way to help teams already implemented OOI with old flow to migrate to new one. Moreover, we need to gather more feedback and understand if we should to invest more time to server implementation on current stage.

Basically here we just generating `statics/editor` and `statics/settings` directories with html and vm content. Then we can just take it's url from CDN (like a lot of teams currently doing) and include in dev center as an editor and settings apps.

Here, we also fixed css urls, since it was hardcoded to localhost and had no chance to work on prod.

### CODE REVIEW

File: packages/yoshi-flow-editor-runtime/src/WidgetWrapper.tsx [hunk 1]
```diff
@@ -24,13 +25,13 @@ const PublicDataProvider: typeof React.Component =
...
 
-const WidgetWrapper = (UserComponent: typeof React.Component) => (
+const WidgetWrapper = (UserComponent: typeof React.Component, name: string) => (
   props: IHostProps & IFrameworkProps & IControllerContext,
 ) => {
   return (
     <div>
       <link
-        href="https://localhost:3200/todoViewerWidget.css"
+        href={`${window.__STATICS_BASE_URL__}${name}ViewerWidget.css`}
```
Comment (yavorsky) [inline packages/yoshi-flow-editor-runtime/src/WidgetWrapper.tsx]: I'll make a better solution for css files, it's just a quick fix better than hardcoded localhist :/

File: packages/yoshi-flow-editor/src/server/vm.ts [hunk 2]
```diff
@@ -3,9 +3,21 @@ import { Engine } from 'velocity';
...
+    >;
+  } = {},
+) =>
   new Engine({ template: url }).render({
     ...velocityData,
     ...velocityDataPrivate,
     ...data,
+    clientTopology: {
+      ...(velocityData.clientTopology ? velocityData.clientTopology : {}),
+      ...(data.clientTopology ? data.clientTopology : {}),
```
Comment (yavorsky) [inline packages/yoshi-flow-editor/src/server/vm.ts]: We need this to merge nested objects.

File: packages/yoshi-flow-editor/src/commands/htmlGenerator.ts [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+import { STATICS_DIR } from 'yoshi-config/build/paths';
+import renderVM from '../server/vm';
+import { FlowEditorModel } from '../model';
+
+type vmType = 'editor' | 'settings';
+
+const templatesPath = path.resolve(__dirname, '../server/templates');
+
+const vmPaths: Record<vmType, string> = {
+  editor: path.join(templatesPath, './editorApp.vm'),
```
Comment (bordi) [inline packages/yoshi-flow-editor/src/commands/htmlGenerator.ts]: **TBD as part of https://github.com/wix/yoshi/issues/2013**
There will be more templates in few cases:
- separate Settings Panel for mobile
- separate Editor App for mobile

Hence, should be glob pattern be used here?

File: packages/yoshi-flow-editor/src/commands/htmlGenerator.ts [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    clientTopology: {
+      staticsBaseUrl: '../',
+    },
+  });
+  fs.copyFileSync(vmPath, path.join(destinationDir, `${widgetName}.vm`));
+  fs.writeFileSync(path.join(destinationDir, `${widgetName}.html`), rendered);
+};
+
+const getOutputFilename = (type: vmType) => {
+  return path.resolve(path.join(STATICS_DIR, type));
```
Comment (ranyitz) [inline packages/yoshi-flow-editor/src/commands/htmlGenerator.ts]: Easier to read as
```suggestion
  return path.join(process.cwd(), STATICS_DIR, type);
```

File: packages/yoshi-flow-editor/src/commands/htmlGenerator.ts [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+};
+
+const generateHTML = (type: vmType, widgetName: string) => {
+  generateHTMLFromVM(getOutputFilename(type), vmPaths[type], {
```
Comment (ranyitz) [inline packages/yoshi-flow-editor/src/commands/htmlGenerator.ts]: `getOutputFilename` could be better named `getOutputDirectory`

But maybe it could just be an implementation detail of `generateHTMLFromVM`

File: packages/yoshi-flow-editor/src/commands/htmlGenerator.ts [hunk 6]
```diff
@@ -0,0 +1,52 @@
+import path from 'path';
+import fs from 'fs-extra';
+import { STATICS_DIR } from 'yoshi-config/build/paths';
+import renderVM from '../server/vm';
+import { FlowEditorModel } from '../model';
+
+type vmType = 'editor' | 'settings';
+
+const templatesPath = path.resolve(__dirname, '../server/templates');
```
Comment (ranyitz) [inline packages/yoshi-flow-editor/src/commands/htmlGenerator.ts]: `path.resolve(arg)` is `path.join(process.cwd(), arg)`. We don't need the `process.cwd()` resolution here because `__dirname` is absolute.

While it's not critical, I would use `path.join` here.

**Comment by sdoomz** [comment]
I like the idea of creating index.html during the build phase and putting it to the statics folder. 
In this case, you don't even need to think about the server and all your client app is loaded from a CDN. 

Pros:   
1. Content is delivered a bit faster. Because of DNS resolution and 0 server processing time. 
2. Reliability. `index.html` is quite stable and doesn't require additional monitoring tools or health checks. 

On the other hand, global variables as `__STATICS_BASE_URL__` are still injected and it's possible to output `index.html` file either with `app.js` or `app.min.js` based on your `NODE_ENV`.

In Search App we used a similar, but less flexible, approach - [just html file](https://github.com/wix-private/site-search/blob/master/packages/search-results-widget/src/editorWidget.html) and it took us some time to implement it. 

Summarizing everything, it would be really nice to have this behaviour implemented in the new `editor-flow` out of the box. 

Thanks Artem.

File: packages/yoshi-flow-editor/src/commands/htmlGenerator.ts [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    widgetName,
+  });
+};
+
+const syncHTMLDirectory = (type: vmType) => {
+  fs.mkdirpSync(path.join(STATICS_DIR, type));
+};
+
+export const generateEditorHTMLFiles = (model: FlowEditorModel) => {
+  syncHTMLDirectory('editor');
```
[the remark this answers is not in the document]
Comment (yavorsky) [reply packages/yoshi-flow-editor/src/commands/htmlGenerator.ts]: It isn't :(

File: packages/yoshi-flow-editor-runtime/src/WidgetWrapper.tsx (hunk 1 above)
Comment (yavorsky) [reply packages/yoshi-flow-editor-runtime/src/WidgetWrapper.tsx]: Not sure I got it. You mean `__webpack_public_path__` ?

**Comment by yavorsky** [comment]
I've added a basic README file, which explain the basic use case.
I'm going to add more docs, with use cases for dev-center registration and point current doc here.

File: packages/yoshi-flow-editor-runtime/src/WidgetWrapper.tsx (hunk 1 above)
Comment (ranyitz) [reply packages/yoshi-flow-editor-runtime/src/WidgetWrapper.tsx]: I mean that we know what would be the public path, and I think that it's equivalent to the statics base url, so we can inject it in build time instead of taking it from the window.

In any case, if it's working right now we can continue using this methodology and update it in a later PR
